examples/seismic/source.py

```diff
@@ -1,60 +1,114 @@
 from devito.function import SparseTimeFunction
 from devito.logger import error
 
+from cached_property import cached_property
+from collections import namedtuple
+
 import numpy as np
 try:
     import matplotlib.pyplot as plt
 except:
     plt = None
 
-__all__ = ['PointSource', 'Receiver', 'Shot', 'RickerSource', 'GaborSource']
+__all__ = ['PointSource', 'Receiver', 'Shot', 'WaveletSource',
+           'RickerSource', 'GaborSource']
 
 
 class PointSource(SparseTimeFunction):
     """Symbolic data object for a set of sparse point sources
 
-    :param name: Name of the symbol representing this source
+    :param name: Name of the symbol representing this source.
     :param grid: :class:`Grid` object defining the computational domain.
-    :param coordinates: Point coordinates for this source
-    :param data: (Optional) Data values to initialise point data
-    :param ntime: (Optional) Number of timesteps for which to allocate data
-    :param npoint: (Optional) Number of sparse points represented by this source
-    :param time_order: (Optional) Time discretization order (defaults to 2)
+    :param coordinates: Point coordinates for this source.
+    :param time_range: (Optional) Tuple of time origin, time interval and
+                       number of timesteps, (t0, dt, ntime). Can only set
+                       one of time_range and time_values.
+    :param time_values: (Optional) Time values. Can only set one of time_range and time_values.
+    :param data: (Optional) Data values to initialise point data.
+    :param npoint: (Optional) Number of sparse points represented by this source.
+    :param time_order: (Optional) Time discretization order (defaults to 2).
     :param dimension: :(Optional) class:`Dimension` object for
-                       representing the number of points in this source
-
-    Note, either the dimensions `ntime` and `npoint` or the fully
-    initialised `data` array need to be provided.
+                       representing the number of points in this source.
     """
 
-    def __new__(cls, name, grid, ntime=None, npoint=None, data=None,
-                coordinates=None, **kwargs):
+    def __new__(cls, name, grid, time_range=None, time_values=None, npoint=None,
+                data=None, coordinates=None, **kwargs):
         p_dim = kwargs.get('dimension', Dimension(name='p_%s' % name))
         time_order = kwargs.get('time_order', 2)
         npoint = npoint or coordinates.shape[0]
-        if data is None:
-            if ntime is None:
-                error('Either data or ntime are required to'
-                      'initialise source/receiver objects')
+
+        if time_range is None and time_values is None:
+            error("Either time_range or time_values must be defined.")
+
+        Time =  namedtuple('Time', ['t0', 'dt', 'nt'])
+        if time_range is None:
+            cls._time_range = Time(time_values[0], time_values[1]-time_values[0], time_values.size)
```

Comment:
I think the clean thing to do here is to create `time_range` as a local variable, then, after ``obj = SparseTimeFunction.__ ...`` you can do ``obj._time_range = time_range``. 

This is, incidentally, the way one subclasses SymPy and NumPy classes